### PR TITLE
test: MPI_REAL16 and MPI_COMPLEX32 may not exist

### PR DESCRIPTION
**Pull request by jeffhammond**
macros to disable MPI_REAL16 and MPI_COMPLEX32.

TODO: needs configure.ac support.  Please help with this.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by raffenet** [comment]
@jeffhammond added a commit that we can squash if it looks good. Just add `Co-authored-by: Ken Raffenetti <raffenet@mcs.anl.gov>`.

**Comment by jeffhammond** [comment]
I trust you on configure stuff. Merge at will. Squash if you like.

File: test/mpi/configure.ac [hunk 1]
```diff
@@ -1578,6 +1578,30 @@ if test "$pac_cv_have_mpi_integer16" = yes ; then
     AC_DEFINE(HAVE_MPI_INTEGER16,1,[Define if MPI_INTEGER16 is available])
 fi
 
+# Check if MPI_REAL16 and MPI_COMPLEX32 are available
+AC_CACHE_CHECK([whether MPI_REAL16 is available],
+pac_cv_have_mpi_real16,[
+    AC_COMPILE_IFELSE([
+        AC_LANG_PROGRAM([#include "mpi.h"],[
+            MPI_Datatype t = MPI_REAL16;
+        ])
```
[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/configure.ac:1587]: This is something an MPI ABI should address

[the remark this answers is not in the document]
Comment (raffenet) [reply test/mpi/configure.ac:1587]: Our tests already handle that case. I guess Jeff in running into instances where the handles are not even defined.

[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/configure.ac:1587]: I see. That makes sense. Can you add a comment that the check does not exclude the case that MPI_REAL16 is MPI_DATATYPE_NULL?

[the remark this answers is not in the document]
Comment (jeffhammond) [reply test/mpi/configure.ac:1587]: Can you initialize to a data type handle if it isn't a compile time constant? I should know this but my brain is full.

[the remark this answers is not in the document]
Comment (hzhou) [reply test/mpi/configure.ac:1587]: The config test is doing that as a local variable in main, so it should work. I would no matter what is the mechanism, the basic requirement is once can declare a static variable initialized with one of the built-in handle. It is a quite common usage, right?

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Review by hzhou — APPROVED** [review]
LGTM

**Pull request merged**